Expose timeout option to Redfish modules

billdodd opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Some operations on some Redfish service implementations/environments may take longer than the default 10 second timeout used in the Ansible `ansible.module_utils.urls.open_url` function.

This change exposes a `timeout` option to allow the operator to override the 10 second default.

Fixes #52332 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
redfish_utils
redfish_facts
redfish_config
redfish_command
idrac_redfish_command

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
**Before**

Given a task like this:

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
  tasks:
    - name: Set BIOS to defaults
      redfish_config:
        category: Systems
        command: SetBiosDefaultSettings
        baseuri: "{{ baseuri }}"
        username: "{{ username }}"
        password: "{{ password }}"
```

If the default timeout is too short for the command on a service, it will get an error like this:

```
TASK [Set BIOS to defaults] ************************************************
fatal: [sys1]: FAILED! => {"changed": false, "msg": "Failed POST operation against Redfish API server: ('The read operation timed out',)"}

PLAY RECAP *********************************************************************
sys1                      : ok=4    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
``` 

**After**

Now the task can be updated like this to override the default timeout and obtain a successful result:

```
  tasks:
    - name: Set BIOS to defaults
      redfish_config:
        category: Systems
        command: SetBiosDefaultSettings
        baseuri: "{{ baseuri }}"
        username: "{{ username }}"
        password: "{{ password }}"
        timeout: 30
```

```
TASK [Set BIOS to defaults] ************************************************
ok: [sys1]

PLAY RECAP *********************************************************************
sys1                      : ok=6    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

billdodd committed:
added timeout option to Redfish modules

jose-delarosa commented:
Pretty straight-forward. Tested it, looks good. Thanks @billdodd.

shipit

gundalow commented on lib/ansible/modules/remote_management/redfish/idrac_redfish_command.py:
```diff
@@ -41,6 +41,12 @@
+    description:
+      - Timeout in seconds for URL requests to OOB controller
+    default: 10
+    version_added: "2.8"
```
```suggestion
    type: int
    version_added: '2.8'
```

gundalow commented on lib/ansible/modules/remote_management/redfish/idrac_redfish_command.py:
```diff
@@ -41,6 +41,12 @@
     description:
       - Password for authentication with OOB controller
+  timeout:
+    required: false
```
This is the default so can be removed.
```suggestion
```

gundalow commented on lib/ansible/modules/remote_management/redfish/redfish_command.py:
```diff
@@ -68,6 +68,12 @@
     description:
       - bootdevice when setting boot configuration
+  timeout:
+    required: false
```
```suggestion
```

gundalow commented on lib/ansible/modules/remote_management/redfish/redfish_command.py:
```diff
@@ -68,6 +68,12 @@
+    description:
+      - Timeout in seconds for URL requests to OOB controller
+    default: 10
+    version_added: "2.8"
```
```suggestion
    type: int
    version_added: '2.8'
```

gundalow commented on lib/ansible/modules/remote_management/redfish/redfish_config.py:
```diff
@@ -55,6 +55,12 @@
     default: 'null'
     version_added: "2.8"
+  timeout:
+    required: false
```
```suggestion
```

gundalow commented on lib/ansible/modules/remote_management/redfish/redfish_config.py:
```diff
@@ -55,6 +55,12 @@
+    description:
+      - Timeout in seconds for URL requests to OOB controller
+    default: 10
+    version_added: "2.8"
```
```suggestion
    type: int
    version_added: "2.8"
```

gundalow commented on lib/ansible/modules/remote_management/redfish/redfish_facts.py:
```diff
@@ -43,6 +43,12 @@
     description:
       - Password for authentication with OOB controller
+  timeout:
+    required: false
```
```suggestion
```

gundalow commented on lib/ansible/modules/remote_management/redfish/redfish_facts.py:
```diff
@@ -43,6 +43,12 @@
+    description:
+      - Timeout in seconds for URL requests to OOB controller
+    default: 10
+    version_added: "2.8"
```
```suggestion
    type: int
    version_added: '2.8'
```

gundalow committed:
Apply suggestions from code review

Removed 'required: false' and added 'type: int' to 'timeout' documentation string.

Co-Authored-By: billdodd <billdodd@gmail.com>

billdodd commented:
@gundalow - Thanks for the documentation improvement suggestions. Applied.

gundalow commented:
Thanks, merged into devel for release in Ansible 2.8
